### ISSUE TITLE
Problem using the code on FAUST Dataset

### ISSUE THREAD

**Issue by daidedou** [kzhou23/shape_pose_disent#3]
Hi,
Thanks for sharing the code!

I'm trying to use the code (encode the shape and pose, then decode) on the FAUST dataset. The meshes doesn't come from a SMPL model (in fact they trained the model as far as I understood), but since they are a deformation of the same template, I thought it should work. 
Here's a gist of what I did https://gist.github.com/daidedou/cb3bded35b3850228c891ada5c7be08d

However, I get some weird results, this is what I get when I load the output mesh in meshlab
![image](https://user-images.githubusercontent.com/46010710/108975807-2a767180-7687-11eb-865e-2ddd748535c0.png)

It probably comes from my configuration, do you have any idea of the possible problem?
Thank you in advance!

**Comment by kzhou23** [comment]
Hi,

I sometimes encountered this sort of weird output and I traced the problem to loading pretrained model. For me loading the model to cpu alway fixes the problem.  Can you have a try and tell me if it works?

**Comment by daidedou** [comment]
Hi, thanks for the answer!
By loading the model on the cpu, you mean doing this:
torch.load("model.pth", map_location="cpu")
Or did you mean something else?  
Other question, maybe it's because the model was trained in Python 2? This would explain why I need to make a weird conversion to load it properly

**Comment by daidedou** [comment]
miscklicked on close issue sorry

**Comment by daidedou** [comment]
I tried torch.load("model.pth", map_location="cpu"), but it didn't solve the problem. Is there any initialisation when loading FAUST dataset, like a rotation or something?

**Comment by kzhou23** [comment]
> Hi, thanks for the answer!
> By loading the model on the cpu, you mean doing this:
> torch.load("model.pth", map_location="cpu")
> Or did you mean something else?
> Other question, maybe it's because the model was trained in Python 2? This would explain why I need to make a weird conversion to load it properly

Yes, I mean you can try loading the pretrained model on cpu and carrying the inference there. The development environment I used is Python 2. But if everything works smoothly in Python 3 in your case, it should not be a problem.

**Comment by kzhou23** [comment]
> I tried torch.load("model.pth", map_location="cpu"), but it didn't solve the problem. Is there any initialisation when loading FAUST dataset, like a rotation or something?

Did you get this corrupted output during training? Or it only fails at inference time?

If it occurs at training time, did you observe training loss and validation loss decreasing?

Did it only happen to pose transfer, or reconstruction fails as well?

**Comment by daidedou** [comment]
> Did you get this corrupted output during training? Or it only fails at inference time?

For now I'm just trying to apply the trained model on 3D models 

> Did it only happen to pose transfer, or reconstruction fails as well?

I'm just doing encode - decode, so we can say that it's the reconstruction that fails

**Comment by daidedou** [comment]
Hi, 
I'm sorry, I can't get to use properly the model. Could you confirm that on the FAUST dataset it works on your side? I tried using it on CPU it fails too. I'm sure I'm doing something wrong.
Another question, in your paper you use the distance between the SMPL parameters of the original and reconstructed body. Do you have code to manage this measure? In particular, how do you convert the pose parameters to quaternions?

**Comment by kzhou23** [comment]
Sorry for this delayed response.

I'm not quite sure how you handled it on your side. Did you train the model on FAUST or did you use the pretrained model? If you ran the training yourself, did you observe the loss converging?

A simple code snippet for converting axis-angle representation to quaterions:

```
def pose2quat(pose):
    pose = pose.reshape(-1, 3)
    thetas = np.linalg.norm(pose, axis=1)
    axes = pose / thetas[:, np.newaxis]
    C = np.cos(thetas / 2)[:, np.newaxis]
    S = axes * np.sin(thetas / 2)[:, np.newaxis]
    quats =  np.concatenate((S, C), axis=1)
    quats[thetas==0] = np.array([[0, 0, 0, 1]])
    return quats
```

**Comment by daidedou** [comment]
Hi!

Thanks for the answer! I was probably not clear. I tried to play with your results by applying the method on FAUST dataset. I tried to use the pretrained model on shapes of the dataset (just for an encode-decode for a first step, no interpolation in the latent space), but it gave me the weird mesh shown on the opening of the issue.

Thank you very much for the code snippet!

**Comment by kzhou23** [comment]
The template of FAUST does have the same number of vertices as SMPL but vertices order also matter in this case. You can take a look at DFaust which is registered to SMPL for sure.
